// @aws-ddd-api/shared/logging/logger.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';

export interface LogOptions {
  event?: APIGatewayProxyEvent & {
    awsRequestId?: string;
    userId?: string;
    userEmail?: string;
    userRole?: string;
  };
  error?: unknown;
  extra?: Record<string, unknown>;
  scope?: string;
}

const SENSITIVE_KEY_PATTERN = /(authorization|cookie|password|secret|token|jwt|api[-_]?key)/i;

function redactValue(key: string, value: unknown, depth = 0): unknown {
  if (SENSITIVE_KEY_PATTERN.test(key)) {
    return '[REDACTED]';
  }

  if (depth > 4 || value === null || value === undefined) {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map((item) => redactValue(key, item, depth + 1));
  }

  if (typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([entryKey, entryValue]) => [
        entryKey,
        redactValue(entryKey, entryValue, depth + 1),
      ])
    );
  }

  return value;
}

function getRequestLogContext(event: LogOptions['event']) {
  if (!event) {
    return undefined;
  }

  return {
    requestId: event.requestContext?.requestId,
    awsRequestId: event.awsRequestId,
    stage: event.requestContext?.stage,
    method: event.httpMethod,
    resource: event.resource,
    path: event.path,
    userId: event.userId,
    userEmail: event.userEmail,
    userRole: event.userRole,
  };
}

export function serializeError(error: unknown) {
  if (!error) {
    return undefined;
  }

  if (error instanceof Error) {
    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
    };
  }

  return redactValue('error', error);
}

function writeStructuredLog(level: 'info' | 'warn' | 'error', message: string, options: LogOptions = {}) {
  const entry = {
    timestamp: new Date().toISOString(),
    level,
    message,
    scope: options.scope,
    request: getRequestLogContext(options.event),
    error: serializeError(options.error),
    extra: options.extra ? redactValue('extra', options.extra) : undefined,
  };

  const logger = level === 'error' ? console.error : level === 'warn' ? console.warn : console.log;
  logger(JSON.stringify(entry));
}

export function logInfo(message: string, options?: LogOptions): void {
  writeStructuredLog('info', message, options);
}

export function logWarn(message: string, options?: LogOptions): void {
  writeStructuredLog('warn', message, options);
}

export function logError(message: string, options?: LogOptions): void {
  writeStructuredLog('error', message, options);
}
