// @aws-ddd-api/shared/http/cors.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';
import type { ResponseHelpers } from './response';

type RuntimeEnv = Record<string, string | undefined>;
type RequestEvent = APIGatewayProxyEvent & { awsRequestId?: string };

const env = (globalThis as unknown as { process: { env: RuntimeEnv } }).process.env;

function getAllowedOrigins(): string[] {
  const value = env.ALLOWED_ORIGINS || '';

  return value
    .split(',')
    .map((origin) => origin.trim())
    .filter(Boolean);
}

function getRequestOrigin(event: RequestEvent): string | null {
  const origin = event.headers?.origin || event.headers?.Origin;
  return typeof origin === 'string' && origin.trim() ? origin.trim() : null;
}

function isProductionStage(): boolean {
  return env.STAGE_NAME === 'production';
}

function isWildcardDevelopment(): boolean {
  return !isProductionStage() && env.ALLOWED_ORIGINS === '*';
}

export function corsHeaders(event: RequestEvent): Record<string, string> {
  const origin = getRequestOrigin(event);

  if (!origin) {
    return {};
  }

  if (isWildcardDevelopment()) {
    return {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Request-Id,x-api-key',
      'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,PATCH,DELETE',
    };
  }

  const isAllowed = getAllowedOrigins().some(
    (allowedOrigin) => allowedOrigin.toLowerCase() === origin.toLowerCase()
  );

  if (!isAllowed) {
    return {};
  }

  return {
    'Access-Control-Allow-Origin': origin,
    'Access-Control-Allow-Credentials': 'true',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Request-Id,x-api-key',
    'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,PATCH,DELETE',
  };
}

export function handleOptions(event: RequestEvent, response: ResponseHelpers) {
  if (event.httpMethod !== 'OPTIONS') {
    return null;
  }

  const headers = corsHeaders(event);

  if (Object.keys(headers).length === 0) {
    return response.errorResponse(403, 'common.originNotAllowed', event, undefined, {
      message: 'CORS preflight origin denied',
      scope: 'shared.http.cors',
    });
  }

  return response.noContentResponse(204, event, headers, {
    message: 'CORS preflight allowed',
    scope: 'shared.http.cors',
  });
}
