// @aws-ddd-api/shared/http/cors.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';

type RuntimeEnv = Record<string, string | undefined>;
type RequestEvent = APIGatewayProxyEvent & { awsRequestId?: string };

const env = (globalThis as unknown as { process: { env: RuntimeEnv } }).process.env;

function getAllowedOrigins(): string[] {
  const value = env.ALLOWED_ORIGINS || '';

  return value
    .split(',')
    .map((origin) => origin.trim())
    .filter(Boolean);
}

function getRequestOrigin(event: RequestEvent): string | null {
  const origin = event.headers?.origin || event.headers?.Origin;
  return typeof origin === 'string' && origin.trim() ? origin.trim() : null;
}

function isProductionStage(): boolean {
  return env.STAGE_NAME === 'production';
}

function isWildcardDevelopment(): boolean {
  return !isProductionStage() && env.ALLOWED_ORIGINS === '*';
}

export function corsHeaders(event: RequestEvent): Record<string, string> {
  const origin = getRequestOrigin(event);

  if (!origin) {
    return {};
  }

  if (isWildcardDevelopment()) {
    return {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Request-Id',
      'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,PATCH,DELETE',
    };
  }

  const isAllowed = getAllowedOrigins().some(
    (allowedOrigin) => allowedOrigin.toLowerCase() === origin.toLowerCase()
  );

  if (!isAllowed) {
    return {};
  }

  return {
    'Access-Control-Allow-Origin': origin,
    'Access-Control-Allow-Credentials': 'true',
    'Access-Control-Allow-Headers': 'Content-Type,Authorization,X-Request-Id',
    'Access-Control-Allow-Methods': 'OPTIONS,GET,POST,PUT,PATCH,DELETE',
  };
}

export function handleOptions(event: RequestEvent) {
  if (event.httpMethod !== 'OPTIONS') {
    return null;
  }

  const headers = corsHeaders(event);

  if (Object.keys(headers).length === 0) {
    return {
      statusCode: 403,
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        success: false,
        errorKey: 'common.originNotAllowed',
        error: 'common.originNotAllowed',
        ...(event.awsRequestId ? { requestId: event.awsRequestId } : {}),
      }),
    };
  }

  return {
    statusCode: 204,
    headers,
    body: '',
  };
}
