// @aws-ddd-api/shared/http/handler.ts
import type { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
import { handleOptions } from './cors';
import { logError } from '../logging/logger';
import type { ResponseHelpers } from './response';

/**
 * Internal route context passed from the API Gateway handler adapter into routers and services.
 */
export interface ApiRouteContext<
  TEvent extends APIGatewayProxyEvent = APIGatewayProxyEvent
> {
  event: TEvent & { awsRequestId?: string };
  body: unknown;
}

/**
 * Function signature for a route dispatcher created by `createRouter`.
 */
export type ApiRouteRequest<
  TEvent extends APIGatewayProxyEvent = APIGatewayProxyEvent
> = (routeContext: ApiRouteContext<TEvent>) => Promise<APIGatewayProxyResult>;

/**
 * Options for the API Gateway handler adapter.
 */
export interface CreateApiGatewayHandlerOptions {
  /** Domain response singleton used for handler-level unhandled error responses. */
  response: ResponseHelpers;
  /** Optional hook called before the default unhandled-error response is returned. */
  onError?: (error: unknown, event: APIGatewayProxyEvent) => void | Promise<void>;
}

/**
 * Maps an error-like value to an HTTP status code.
 *
 * @param error - Any thrown value.
 * @returns A valid 4xx/5xx status code, defaulting to 500.
 */
function getErrorStatusCode(error: unknown): number {
  const statusCode = (error as { statusCode?: unknown })?.statusCode;

  if (typeof statusCode === 'number' && statusCode >= 400 && statusCode <= 599) {
    return statusCode;
  }

  return 500;
}

/**
 * Maps an error-like value to a shared locale key.
 *
 * 5xx errors intentionally hide the original message by default. If a 4xx
 * error message is already a dotted locale key, that key is preserved.
 *
 * @param error - Any thrown value.
 * @param statusCode - Status code already derived for the error.
 * @returns Shared locale key or raw 4xx message fallback.
 */
function getErrorKey(error: unknown, statusCode: number): string {
  const errorKey = (error as { errorKey?: unknown })?.errorKey;
  if (typeof errorKey === 'string' && errorKey.trim() && statusCode < 500) {
    return errorKey;
  }

  const message = (error as { message?: unknown })?.message;

  if (typeof message === 'string' && message.includes('.') && statusCode < 500) {
    return message;
  }

  if (statusCode === 401) return 'common.unauthorized';
  if (statusCode === 403) return 'common.forbidden';
  if (statusCode === 404) return 'common.notFound';
  if (statusCode === 405) return 'common.methodNotAllowed';
  if (statusCode === 429) return 'common.rateLimited';
  if (statusCode === 503) return 'common.serviceUnavailable';

  return statusCode >= 500 ? 'common.internalError' : 'common.badRequest';
}

/**
 * Safely parses a JSON string request body for route context.
 *
 * Invalid JSON returns the original string so downstream validation can decide
 * whether to reject it as malformed input.
 *
 * @param body - Raw API Gateway `event.body`.
 * @returns Parsed JSON, `null` for empty bodies, or the original string on parse failure.
 */
function safeJsonParse(body: string | null | undefined): unknown {
  if (!body) {
    return null;
  }

  try {
    return JSON.parse(body);
  } catch {
    return body;
  }
}

/**
 * Adapts an internal route dispatcher into a real AWS Lambda handler.
 *
 * The returned handler sets Lambda runtime options, attaches `awsRequestId` to
 * the event, parses `event.body`, and calls `routeRequest({ event, body })`.
 *
 * @param routeRequest - Internal route dispatcher, usually returned by `createRouter`.
 * @param options - Optional response singleton and error handling behavior.
 * @returns Lambda handler compatible with API Gateway proxy events.
 */
export function createApiGatewayHandler(
  routeRequest: ApiRouteRequest,
  options: CreateApiGatewayHandlerOptions
) {
  const { response } = options;

  return async function handler(
    event: APIGatewayProxyEvent,
    context: Context
  ): Promise<APIGatewayProxyResult> {
    context.callbackWaitsForEmptyEventLoop = false;
    const requestEvent = event as APIGatewayProxyEvent & { awsRequestId?: string };
    requestEvent.awsRequestId = context.awsRequestId;

    try {
      const optionsResponse = handleOptions(requestEvent);
      if (optionsResponse) {
        return optionsResponse;
      }

      return await routeRequest({
        event: requestEvent,
        body: safeJsonParse(event.body),
      });
    } catch (error) {
      await options.onError?.(error, event);

      logError('Unhandled API request error', {
        event: requestEvent,
        error,
        scope: 'shared.http.handler',
      });

      const statusCode = getErrorStatusCode(error);
      return response.errorResponse(statusCode, getErrorKey(error, statusCode), requestEvent);
    }
  };
}
