// @aws-ddd-api/shared/http/handler.ts
import type { APIGatewayProxyEvent, APIGatewayProxyResult, Context } from 'aws-lambda';
import { logError } from '../logging/logger';
import { errorJson, json, safeJsonParse } from './response';

export type JsonResponder = typeof json;

export interface ApiRouteContext<
  TEvent extends APIGatewayProxyEvent = APIGatewayProxyEvent
> {
  event: TEvent & { awsRequestId?: string };
  body: unknown;
  json: JsonResponder;
}

export type ApiRouteRequest<
  TEvent extends APIGatewayProxyEvent = APIGatewayProxyEvent
> = (routeContext: ApiRouteContext<TEvent>) => Promise<APIGatewayProxyResult>;

export interface CreateApiGatewayHandlerOptions {
  includeErrorDetails?: boolean;
  onError?: (error: unknown, event: APIGatewayProxyEvent) => void | Promise<void>;
}

function getErrorStatusCode(error: unknown): number {
  const statusCode = (error as { statusCode?: unknown })?.statusCode;

  if (typeof statusCode === 'number' && statusCode >= 400 && statusCode <= 599) {
    return statusCode;
  }

  return 500;
}

function getErrorMessage(error: unknown, statusCode: number): string {
  const message = (error as { message?: unknown })?.message;

  if (typeof message === 'string' && message.trim() && statusCode < 500) {
    return message;
  }

  return statusCode >= 500 ? 'Internal server error' : 'Request failed';
}

export function createApiGatewayHandler(
  routeRequest: ApiRouteRequest,
  options: CreateApiGatewayHandlerOptions = {}
) {
  return async function handler(
    event: APIGatewayProxyEvent,
    context: Context
  ): Promise<APIGatewayProxyResult> {
    context.callbackWaitsForEmptyEventLoop = false;
    const requestEvent = event as APIGatewayProxyEvent & { awsRequestId?: string };
    requestEvent.awsRequestId = context.awsRequestId;

    try {
      return await routeRequest({
        event: requestEvent,
        body: safeJsonParse(event.body),
        json,
      });
    } catch (error) {
      await options.onError?.(error, event);

      logError('Unhandled API request error', {
        event: requestEvent,
        error,
        scope: 'shared.http.handler',
      });

      const statusCode = getErrorStatusCode(error);
      return errorJson(statusCode, getErrorMessage(error, statusCode), {
        details: options.includeErrorDetails ? String((error as { stack?: unknown })?.stack || error) : undefined,
        requestId: context.awsRequestId,
      });
    }
  };
}
