// @aws-ddd-api/shared/http/response.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';
import { getRequestLocale, getTranslation, loadTranslations, type TranslationDictionaries } from '../i18n';

/**
 * API Gateway proxy JSON response shape returned by shared response helpers.
 */
export interface JsonResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

/**
 * Common response options that are not tied to a specific API Gateway event.
 */
interface ResponseOptions {
  domainTranslations?: TranslationDictionaries;
  extraHeaders?: Record<string, string>;
  locale?: string;
  requestId?: string;
}

/**
 * Response options that can infer locale and request id from an API Gateway event.
 */
interface EventResponseOptions extends ResponseOptions {
  event?: APIGatewayProxyEvent & { awsRequestId?: string };
}

/**
 * Configuration for a response helper singleton.
 */
export interface CreateResponseOptions {
  domainTranslations?: TranslationDictionaries;
}

/**
 * Response helper singleton returned by `createResponse`.
 */
export interface ResponseHelpers {
  errorResponse: (
    statusCode: number,
    errorKey: string,
    event: APIGatewayProxyEvent & { awsRequestId?: string },
    extraHeaders?: Record<string, string>
  ) => JsonResponse;
  successResponse: (
    statusCode: number,
    event: APIGatewayProxyEvent & { awsRequestId?: string },
    data?: Record<string, unknown>,
    extraHeaders?: Record<string, string>
  ) => JsonResponse;
}

/**
 * Serializes a payload as an API Gateway JSON response.
 *
 * This is the low-level response helper. It does not add `success`, translate
 * messages, or infer locale.
 *
 * @param statusCode - HTTP status code.
 * @param payload - JSON-serializable response body.
 * @param extraHeaders - Optional headers merged after `content-type`.
 * @returns API Gateway proxy response.
 */
function json(
  statusCode: number,
  payload: unknown,
  extraHeaders: Record<string, string> = {}
): JsonResponse {
  return {
    statusCode,
    headers: {
      'content-type': 'application/json',
      ...extraHeaders,
    },
    body: JSON.stringify(payload),
  };
}

/**
 * Resolves the locale to use for a response.
 *
 * @param options - Response options that may include an explicit locale or event.
 * @returns Locale string used for translation lookup.
 */
function getResponseLocale(options: EventResponseOptions): string {
  if (options.locale) {
    return options.locale;
  }

  if (options.event) {
    return getRequestLocale(options.event);
  }

  return 'en';
}

/**
 * Resolves a request id from explicit options or an API Gateway event.
 *
 * @param options - Response options that may include `requestId` or `event`.
 * @returns Request id when available.
 */
function getResponseRequestId(options: EventResponseOptions): string | undefined {
  return options.requestId || options.event?.awsRequestId || options.event?.requestContext?.requestId;
}

/**
 * Translates a locale key when the message is a string.
 *
 * Unknown keys fall back to the input message, matching Stage 1 behavior.
 *
 * @param message - Translation key or raw message.
 * @param locale - Locale used for common translation lookup.
 * @returns Translated message or original value.
 */
function translateMessage(message: unknown, locale: string, domainTranslations?: TranslationDictionaries): unknown {
  if (typeof message !== 'string') {
    return message;
  }

  return getTranslation(loadTranslations(locale, domainTranslations), message, message);
}

/**
 * Creates a Stage 1-compatible translated error response.
 *
 * Response body shape is `{ success: false, errorKey, error, requestId? }`.
 *
 * @param statusCode - HTTP status code.
 * @param errorKey - Locale key such as `common.unauthorized`, or a raw fallback message.
 * @param options - Optional event, explicit locale, request id, and headers.
 * @returns API Gateway proxy response.
 */
function errorResponse(
  statusCode: number,
  errorKey: string,
  options: EventResponseOptions = {}
): JsonResponse {
  const locale = getResponseLocale(options);
  const requestId = getResponseRequestId(options);

  return json(
    statusCode,
    {
      success: false,
      errorKey,
      error: translateMessage(errorKey, locale, options.domainTranslations),
      ...(requestId ? { requestId } : {}),
    },
    options.extraHeaders
  );
}

/**
 * Creates a Stage 1-compatible success response.
 *
 * If `data.message` is a string locale key, it is translated with the request
 * locale. Other data fields are passed through unchanged.
 *
 * @param statusCode - HTTP status code.
 * @param data - Additional response body fields merged after `success: true`.
 * @param options - Optional event, explicit locale, request id, and headers.
 * @returns API Gateway proxy response.
 */
function successResponse(
  statusCode: number,
  data: Record<string, unknown> = {},
  options: EventResponseOptions = {}
): JsonResponse {
  const locale = getResponseLocale(options);
  const requestId = getResponseRequestId(options);
  const responseData = { ...data };

  if (typeof responseData.message === 'string') {
    responseData.message = translateMessage(responseData.message, locale, options.domainTranslations);
  }

  return json(
    statusCode,
    {
      success: true,
      ...responseData,
      ...(requestId ? { requestId } : {}),
    },
    options.extraHeaders
  );
}

/**
 * Creates response helpers bound to optional domain locale dictionaries.
 *
 * Use this once per Lambda/domain and export the returned singleton from a
 * local `src/utils/response.ts` module.
 *
 * @param options - Optional domain locale dictionaries.
 * @returns Standard error and success response helpers.
 */
export function createResponse(options: CreateResponseOptions = {}): ResponseHelpers {
  return {
    errorResponse(statusCode, errorKey, event, extraHeaders) {
      return errorResponse(statusCode, errorKey, {
        domainTranslations: options.domainTranslations,
        event,
        extraHeaders,
      });
    },
    successResponse(statusCode, event, data = {}, extraHeaders) {
      return successResponse(statusCode, data, {
        domainTranslations: options.domainTranslations,
        event,
        extraHeaders,
      });
    },
  };
}
