// @aws-ddd-api/shared/http/response.ts
import type { APIGatewayProxyResult } from 'aws-lambda';

export interface JsonResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

export interface ErrorJsonOptions {
  code?: string;
  details?: unknown;
  extraHeaders?: Record<string, string>;
  requestId?: string;
}

export interface CorsHeaderOptions {
  origin?: string | null;
  allowedOrigins?: string[];
  allowCredentials?: boolean;
  allowHeaders?: string;
  allowMethods?: string;
}

export function json(
  statusCode: number,
  payload: unknown,
  extraHeaders: Record<string, string> = {}
): JsonResponse {
  return {
    statusCode,
    headers: {
      'content-type': 'application/json',
      ...extraHeaders,
    },
    body: JSON.stringify(payload),
  };
}

export function errorJson(
  statusCode: number,
  message = 'Request failed',
  options: ErrorJsonOptions = {}
): JsonResponse {
  return json(
    statusCode,
    {
      success: false,
      error: {
        message,
        ...(options.code ? { code: options.code } : {}),
        ...(options.requestId ? { requestId: options.requestId } : {}),
        ...(options.details !== undefined ? { details: options.details } : {}),
      },
    },
    options.extraHeaders
  );
}

export function safeJsonParse(body: string | null | undefined): unknown {
  if (!body) {
    return null;
  }

  try {
    return JSON.parse(body);
  } catch {
    return body;
  }
}

export function withCorsHeaders<T extends APIGatewayProxyResult>(
  response: T,
  options: CorsHeaderOptions = {}
): T {
  const origin = options.origin?.trim();
  const allowedOrigins = options.allowedOrigins || [];
  const allowsAnyOrigin = allowedOrigins.includes('*');
  const isAllowedOrigin = Boolean(
    origin &&
      (allowsAnyOrigin ||
        allowedOrigins.some((allowedOrigin) => allowedOrigin.toLowerCase() === origin.toLowerCase()))
  );

  if (!isAllowedOrigin && !allowsAnyOrigin) {
    return response;
  }

  return {
    ...response,
    headers: {
      ...response.headers,
      'Access-Control-Allow-Origin': origin || '*',
      'Access-Control-Allow-Headers': options.allowHeaders || 'Content-Type,Authorization,X-Request-Id',
      'Access-Control-Allow-Methods': options.allowMethods || 'OPTIONS,GET,POST,PUT,PATCH,DELETE',
      ...(options.allowCredentials ? { 'Access-Control-Allow-Credentials': 'true' } : {}),
    },
  };
}
