// @aws-ddd-api/shared/http/router.ts
import type { APIGatewayProxyResult } from 'aws-lambda';
import type { ApiRouteContext } from './handler';
import type { ResponseHelpers } from './response';

/**
 * Handler for one matched route.
 */
export type RouteHandler = (routeContext: ApiRouteContext) => Promise<APIGatewayProxyResult>;

/**
 * Route map keyed by API Gateway resource path, raw path, or `METHOD /path`.
 */
export type RouteMap = Record<string, RouteHandler>;

/**
 * Options for the shared route dispatcher.
 */
export interface CreateRouterOptions {
  response: ResponseHelpers;
  notFoundErrorKey?: string;
  methodNotAllowedErrorKey?: string;
}

/**
 * Builds route lookup keys from the incoming API Gateway event.
 *
 * @param routeContext - Current request route context.
 * @returns Candidate keys in lookup priority order.
 */
function getEventRouteCandidates(routeContext: ApiRouteContext): string[] {
  const method = routeContext.event.httpMethod;
  const resource = routeContext.event.resource;
  const path = routeContext.event.path;

  return [
    `${method} ${resource}`,
    `${method} ${path}`,
    resource,
    path,
  ].filter((candidate): candidate is string => Boolean(candidate));
}

/**
 * Detects whether the path exists for another HTTP method.
 *
 * @param routes - Configured route map.
 * @param routeContext - Current request route context.
 * @returns True when a `METHOD /path` route exists for this path but not this method.
 */
function hasRouteForDifferentMethod(routes: RouteMap, routeContext: ApiRouteContext): boolean {
  const resourceOrPath = new Set([routeContext.event.resource, routeContext.event.path].filter(Boolean));

  return Object.keys(routes).some((routeKey) => {
    const separatorIndex = routeKey.indexOf(' ');
    if (separatorIndex === -1) {
      return false;
    }

    return resourceOrPath.has(routeKey.slice(separatorIndex + 1));
  });
}

/**
 * Creates a route dispatcher from a route map.
 *
 * The dispatcher supports API Gateway `event.resource`, raw `event.path`, and
 * method-qualified route keys such as `GET /pet/profile`.
 *
 * @param routes - Route map for one Lambda/domain.
 * @param options - Optional response singleton and messages for 404 and 405 responses.
 * @returns A route dispatcher suitable for `createApiGatewayHandler`.
 */
export function createRouter(routes: RouteMap, options: CreateRouterOptions) {
  const { response } = options;

  return async function routeRequest(routeContext: ApiRouteContext): Promise<APIGatewayProxyResult> {
    const routeAction = getEventRouteCandidates(routeContext)
      .map((candidate) => routes[candidate])
      .find(Boolean);

    if (routeAction) {
      return routeAction(routeContext);
    }

    const routeKey = `${routeContext.event.httpMethod} ${routeContext.event.resource}`;
    const isKnownRoute = hasRouteForDifferentMethod(routes, routeContext);

    return response.errorResponse(
      isKnownRoute ? 405 : 404,
      isKnownRoute
        ? options.methodNotAllowedErrorKey || 'common.methodNotAllowed'
        : options.notFoundErrorKey || 'common.routeNotFound',
      routeContext.event,
      {
        'x-route-key': routeKey,
      }
    );
  };
}
