// @aws-ddd-api/shared/http/router.ts
import type { APIGatewayProxyResult } from 'aws-lambda';
import type { ApiRouteContext } from './handler';

export type RouteHandler = (routeContext: ApiRouteContext) => Promise<APIGatewayProxyResult>;
export type RouteMap = Record<string, RouteHandler>;

export interface CreateRouterOptions {
  notFoundMessage?: string;
  methodNotAllowedMessage?: string;
}

function getEventRouteCandidates(routeContext: ApiRouteContext): string[] {
  const method = routeContext.event.httpMethod;
  const resource = routeContext.event.resource;
  const path = routeContext.event.path;

  return [
    `${method} ${resource}`,
    `${method} ${path}`,
    resource,
    path,
  ].filter((candidate): candidate is string => Boolean(candidate));
}

function hasRouteForDifferentMethod(routes: RouteMap, routeContext: ApiRouteContext): boolean {
  const resourceOrPath = new Set([routeContext.event.resource, routeContext.event.path].filter(Boolean));

  return Object.keys(routes).some((routeKey) => {
    const separatorIndex = routeKey.indexOf(' ');
    if (separatorIndex === -1) {
      return false;
    }

    return resourceOrPath.has(routeKey.slice(separatorIndex + 1));
  });
}

export function createRouter(routes: RouteMap, options: CreateRouterOptions = {}) {
  return async function routeRequest(routeContext: ApiRouteContext): Promise<APIGatewayProxyResult> {
    const routeAction = getEventRouteCandidates(routeContext)
      .map((candidate) => routes[candidate])
      .find(Boolean);

    if (routeAction) {
      return routeAction(routeContext);
    }

    const routeKey = `${routeContext.event.httpMethod} ${routeContext.event.resource}`;
    const isKnownRoute = hasRouteForDifferentMethod(routes, routeContext);

    return routeContext.json(isKnownRoute ? 405 : 404, {
      success: false,
      message: isKnownRoute
        ? options.methodNotAllowedMessage || 'Method not allowed'
        : options.notFoundMessage || 'Route not found',
      routeKey,
    });
  };
}
