// @aws-ddd-api/shared/http/router.ts
import type { APIGatewayProxyResult } from 'aws-lambda';
import type { ApiRouteContext } from './handler';
import type { ResponseHelpers } from './response';
import { logWarn } from '../logging/logger';

/**
 * Handler for one matched route.
 */
export type RouteHandler = (routeContext: ApiRouteContext) => Promise<APIGatewayProxyResult>;

/**
 * Deferred loader for a route handler — a 0-argument function returning a
 * Promise that resolves to a RouteHandler.
 *
 * `createRouter` invokes the loader at most once per execution environment
 * (the resolved handler is cached for subsequent invocations), so the
 * underlying module — and any module-level side effects such as SDK client
 * construction — is only paid for when the route is actually hit.
 */
export type LazyRouteLoader = () => Promise<RouteHandler>;

/**
 * Route map keyed by API Gateway resource path, raw path, or `METHOD /path`.
 *
 * All entries are lazy loaders. To register an already-resolved handler use
 * `() => Promise.resolve(handler)`.
 */
export type RouteMap = Record<string, LazyRouteLoader>;

/**
 * Options for the shared route dispatcher.
 */
export interface CreateRouterOptions {
  response: ResponseHelpers;
  notFoundErrorKey?: string;
  methodNotAllowedErrorKey?: string;
}

/**
 * Builds route lookup keys from the incoming API Gateway event.
 *
 * @param routeContext - Current request route context.
 * @returns Candidate keys in lookup priority order.
 */
function getEventRouteCandidates(routeContext: ApiRouteContext): string[] {
  const method = routeContext.event.httpMethod;
  const resource = routeContext.event.resource;
  const path = routeContext.event.path;

  return [
    `${method} ${resource}`,
    `${method} ${path}`,
    resource,
    path,
  ].filter((candidate): candidate is string => Boolean(candidate));
}

/**
 * Detects whether the path exists for another HTTP method.
 *
 * @param routes - Configured route map.
 * @param routeContext - Current request route context.
 * @returns True when a `METHOD /path` route exists for this path but not this method.
 */
function hasRouteForDifferentMethod(routes: RouteMap, routeContext: ApiRouteContext): boolean {
  const resourceOrPath = new Set([routeContext.event.resource, routeContext.event.path].filter(Boolean));

  return Object.keys(routes).some((routeKey) => {
    const separatorIndex = routeKey.indexOf(' ');
    if (separatorIndex === -1) {
      return false;
    }

    return resourceOrPath.has(routeKey.slice(separatorIndex + 1));
  });
}

/**
 * Creates a route dispatcher from a route map.
 *
 * The dispatcher supports API Gateway `event.resource`, raw `event.path`, and
 * method-qualified route keys such as `GET /pet/profile`.
 *
 * Each route entry is a lazy loader that is invoked at most once per
 * execution environment. The resolved `RouteHandler` is cached so subsequent
 * (warm) invocations on the same path skip the import altogether.
 *
 * @param routes - Route map for one Lambda/domain.
 * @param options - Optional response singleton and messages for 404 and 405 responses.
 * @returns A route dispatcher suitable for `createApiGatewayHandler`.
 */
export function createRouter(routes: RouteMap, options: CreateRouterOptions) {
  const { response } = options;
  const scope = 'shared.http.router';
  // Cache resolved handlers across warm invocations of the same container so
  // each lazy loader runs at most once per execution environment.
  const resolvedHandlers = new Map<string, RouteHandler>();

  async function resolveHandler(
    routeKey: string,
    loader: LazyRouteLoader,
    routeContext: ApiRouteContext
  ): Promise<RouteHandler> {
    const cached = resolvedHandlers.get(routeKey);
    if (cached) {
      logWarn('Router handler resolution', {
        scope,
        event: routeContext.event,
        extra: {
          phase: 'resolveHandler',
          routeKey,
          cacheHit: true,
          durationMs: 0,
        },
      });
      return cached;
    }

    const loaderStartedAt = Date.now();
    const handler = await loader();
    const loaderDurationMs = Date.now() - loaderStartedAt;
    resolvedHandlers.set(routeKey, handler);

    logWarn('Router handler resolution', {
      scope,
      event: routeContext.event,
      extra: {
        phase: 'resolveHandler',
        routeKey,
        cacheHit: false,
        durationMs: loaderDurationMs,
      },
    });

    return handler;
  }

  return async function routeRequest(routeContext: ApiRouteContext): Promise<APIGatewayProxyResult> {
    const routeStartedAt = Date.now();
    let matchedKey: string | undefined;
    let matchedLoader: LazyRouteLoader | undefined;
    for (const candidate of getEventRouteCandidates(routeContext)) {
      const loader = routes[candidate];
      if (loader) {
        matchedKey = candidate;
        matchedLoader = loader;
        break;
      }
    }

    if (matchedKey !== undefined && matchedLoader !== undefined) {
      const resolveStartedAt = Date.now();
      const handler = await resolveHandler(matchedKey, matchedLoader, routeContext);
      const resolveDurationMs = Date.now() - resolveStartedAt;
      logWarn('Router dispatch timing', {
        scope,
        event: routeContext.event,
        extra: {
          phase: 'resolveRouteHandler',
          routeKey: matchedKey,
          durationMs: resolveDurationMs,
        },
      });

      const handlerStartedAt = Date.now();
      const result = await handler(routeContext);
      const handlerDurationMs = Date.now() - handlerStartedAt;
      logWarn('Router dispatch completed', {
        scope,
        event: routeContext.event,
        extra: {
          phase: 'invokeRouteHandler',
          routeKey: matchedKey,
          handlerDurationMs,
          totalDurationMs: Date.now() - routeStartedAt,
        },
      });
      return result;
    }

    const routeKey = `${routeContext.event.httpMethod} ${routeContext.event.resource}`;
    const isKnownRoute = hasRouteForDifferentMethod(routes, routeContext);

    return response.errorResponse(
      isKnownRoute ? 405 : 404,
      isKnownRoute
        ? options.methodNotAllowedErrorKey || 'common.methodNotAllowed'
        : options.notFoundErrorKey || 'common.routeNotFound',
      routeContext.event,
      {
        'x-route-key': routeKey,
      }
    );
  };
}
