// @aws-ddd-api/shared/rate-limit/mongo.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';
import type { Model, Mongoose, Schema } from 'mongoose';

declare const require: (moduleName: string) => {
  createHash: (algorithm: string) => {
    digest: (encoding: 'hex') => string;
    update: (data: string) => ReturnType<ReturnType<typeof require>['createHash']>;
  };
};

const { createHash } = require('crypto');

/**
 * MongoDB rate-limit document shape.
 */
interface MongoRateLimitDocument {
  action: string;
  key: string;
  windowStart: Date;
  count: number;
  expireAt: Date;
}

/**
 * Mongoose model used by the shared Mongo rate limiter.
 */
type MongoRateLimitModel = Model<MongoRateLimitDocument>;

/**
 * Internal options for creating or retrieving the shared rate-limit model.
 */
interface GetMongoRateLimitModelOptions {
  collectionName?: string;
  modelName?: string;
}

/**
 * Internal options for hashing rate-limit key material before storing it.
 */
interface HashRateLimitKeyOptions {
  salt?: string;
}

/**
 * Internal options for building a normalized rate-limit key.
 */
interface CreateRateLimitKeyOptions extends HashRateLimitKeyOptions {
  hash?: boolean;
  separator?: string;
}

/**
 * Result returned after consuming one rate-limit unit.
 */
interface RateLimitResult {
  action: string;
  allowed: boolean;
  count: number;
  limit: number;
  remaining: number;
  resetAt: string;
  retryAfterSeconds: number;
  windowStart: string;
}

/**
 * Internal options for the low-level Mongo counter operation.
 */
interface ConsumeMongoRateLimitOptions {
  action: string;
  failOpen?: boolean;
  key: string;
  limit: number;
  model: MongoRateLimitModel;
  nowMs?: number;
  ttlWindowMultiplier?: number;
  windowSeconds: number;
}

/**
 * Scopes supported by the layered rate limiter. The scope determines which
 * caller attributes are combined to form the rate-limit storage key.
 *
 * - `ip`            — caller IP only (catches single-host abuse across victims)
 * - `identifier`    — supplied identifier only (catches distributed-IP abuse against one victim)
 * - `ip+identifier` — composite of caller IP and identifier (default narrow lane)
 * - `account`       — supplied accountId only (per-authenticated-account lane)
 * - `global`        — no key parts; one bucket per action (last-resort cap)
 */
type RateLimitScope = 'ip' | 'identifier' | 'ip+identifier' | 'account' | 'global';

/**
 * One lane in a layered rate-limit policy set. The request is rejected if
 * any policy is over its limit.
 */
interface RateLimitPolicy {
  limit: number;
  scope: RateLimitScope;
  windowSeconds: number;
}

/**
 * Options for event-based rate limiting.
 */
interface EnforceMongoRateLimitOptions {
  accountId?: string | number | null;
  action: string;
  event: APIGatewayProxyEvent;
  failOpen?: boolean;
  hashKey?: boolean;
  identifier?: string | number | null;
  includeIp?: boolean;
  keySalt?: string;
  limit?: number;
  model: MongoRateLimitModel;
  nowMs?: number;
  policies?: RateLimitPolicy[];
  ttlWindowMultiplier?: number;
  windowSeconds?: number;
}

/**
 * Options for requiring a Mongo-backed rate limit for the current request.
 */
interface RequireMongoRateLimitOptions extends Omit<EnforceMongoRateLimitOptions, 'model'> {
  collectionName?: string;
  modelName?: string;
  mongoose: Mongoose;
}

/**
 * Options for failure-only counters and cooldown checks.
 *
 * Failure counters are persisted in the same `rate_limits` collection but use
 * a distinct `action` (caller's responsibility — e.g. `auth.login.fail`) so
 * they never collide with normal request rate-limit entries.
 */
interface FailureCounterOptions {
  accountId?: string | number | null;
  action: string;
  collectionName?: string;
  cooldownSeconds: number;
  event: APIGatewayProxyEvent;
  failOpen?: boolean;
  hashKey?: boolean;
  identifier?: string | number | null;
  keySalt?: string;
  modelName?: string;
  mongoose: Mongoose;
  nowMs?: number;
  scope?: RateLimitScope;
  threshold: number;
  ttlWindowMultiplier?: number;
}

/**
 * Internal error thrown by `requireMongoRateLimit` when a request is over limit.
 */
class RateLimitError extends Error {
  errorKey = 'common.rateLimited';
  result: RateLimitResult;
  statusCode = 429;

  constructor(result: RateLimitResult) {
    super('common.rateLimited');
    this.name = 'RateLimitError';
    this.result = result;
  }
}

/**
 * Reads a header from an API Gateway event without depending on header casing.
 *
 * @param event - API Gateway event.
 * @param name - Header name to read.
 * @returns Header value, or undefined when not present.
 */
function getHeaderValue(event: APIGatewayProxyEvent, name: string): string | undefined {
  const target = name.toLowerCase();
  const headers = event.headers || {};

  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() === target && typeof value === 'string') {
      return value;
    }
  }

  return undefined;
}

/**
 * Extracts the client IP from an API Gateway event.
 *
 * Prefers the first `X-Forwarded-For` value and falls back to REST/HTTP API
 * request context source IPs.
 *
 * @param event - API Gateway event.
 * @returns Client IP, or `unknown` when not available.
 */
function getClientIp(event: APIGatewayProxyEvent): string {
  const forwardedFor = getHeaderValue(event, 'x-forwarded-for');
  if (typeof forwardedFor === 'string' && forwardedFor.trim()) {
    return forwardedFor.split(',')[0].trim();
  }

  return event.requestContext?.identity?.sourceIp || 'unknown';
}

/**
 * Floors a timestamp to the start of its fixed rate-limit window.
 *
 * @param nowMs - Current epoch milliseconds.
 * @param windowSeconds - Window length in seconds.
 * @returns Window start date.
 */
function toRateLimitWindowStart(nowMs: number, windowSeconds: number): Date {
  const windowMs = windowSeconds * 1000;
  return new Date(Math.floor(nowMs / windowMs) * windowMs);
}

/**
 * Hashes rate-limit key material for safer persistence.
 *
 * Use this for email, phone, token, IP, or other user-identifying key parts.
 *
 * @param value - Raw key material.
 * @param options - Optional salt.
 * @returns SHA-256 hex digest.
 */
function hashRateLimitKey(value: string, options: HashRateLimitKeyOptions = {}): string {
  return createHash('sha256')
    .update(options.salt || '')
    .update(value)
    .digest('hex');
}

/**
 * Builds a deterministic rate-limit key from multiple key parts.
 *
 * By default the returned key is hashed so raw identifiers are not stored.
 *
 * @param parts - Key parts such as client IP, user id, email, or route name.
 * @param options - Hashing and separator options.
 * @returns Stable key for Mongo rate-limit storage.
 */
function createRateLimitKey(
  parts: Array<string | number | null | undefined>,
  options: CreateRateLimitKeyOptions = {}
): string {
  const separator = options.separator || ':';
  const rawKey = parts
    .map((part) => (part === undefined || part === null || part === '' ? 'anonymous' : String(part).trim()))
    .join(separator);

  return options.hash === false ? rawKey : hashRateLimitKey(rawKey, options);
}

/**
 * Creates the Mongo rate-limit schema.
 *
 * The schema uses a unique fixed-window index and a TTL index on `expireAt`.
 *
 * @param mongoose - Domain Lambda's connected/imported Mongoose instance.
 * @returns Mongoose schema for rate-limit counters.
 */
function createMongoRateLimitSchema(mongoose: Mongoose): Schema<MongoRateLimitDocument> {
  const schema = new mongoose.Schema<MongoRateLimitDocument>(
    {
      action: {
        required: true,
        trim: true,
        type: String,
      },
      key: {
        required: true,
        trim: true,
        type: String,
      },
      windowStart: {
        required: true,
        type: Date,
      },
      count: {
        default: 0,
        required: true,
        type: Number,
      },
      expireAt: {
        required: true,
        type: Date,
      },
    },
    {
      timestamps: true,
    }
  );

  schema.index({ action: 1, key: 1, windowStart: 1 }, { unique: true });
  schema.index({ expireAt: 1 }, { expireAfterSeconds: 0 });

  return schema;
}

/**
 * Gets or registers the Mongo rate-limit model on a domain Mongoose instance.
 *
 * Call this after the Lambda's normal DB setup has imported/connected
 * Mongoose.
 *
 * @param mongoose - Domain Lambda's Mongoose instance.
 * @param options - Optional model and collection names.
 * @returns Mongoose model for rate-limit counters.
 */
function getMongoRateLimitModel(
  mongoose: Mongoose,
  options: GetMongoRateLimitModelOptions = {}
): MongoRateLimitModel {
  const modelName = options.modelName || 'RateLimit';
  const collectionName = options.collectionName || 'rate_limits';
  const existingModel = mongoose.models[modelName] as MongoRateLimitModel | undefined;

  return existingModel || mongoose.model<MongoRateLimitDocument>(
    modelName,
    createMongoRateLimitSchema(mongoose),
    collectionName
  );
}

function getRetryAfterSeconds(nowMs: number, resetAt: Date): number {
  return Math.max(0, Math.ceil((resetAt.getTime() - nowMs) / 1000));
}

function createResult(
  action: string,
  count: number,
  limit: number,
  windowStart: Date,
  resetAt: Date,
  nowMs: number
): RateLimitResult {
  return {
    action,
    allowed: count <= limit,
    count,
    limit,
    remaining: Math.max(0, limit - count),
    resetAt: resetAt.toISOString(),
    retryAfterSeconds: getRetryAfterSeconds(nowMs, resetAt),
    windowStart: windowStart.toISOString(),
  };
}

function isDuplicateKeyError(error: unknown): boolean {
  return (error as { code?: unknown })?.code === 11000;
}

function assertPositiveInteger(value: number, name: string): void {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer`);
  }
}

/**
 * Atomically consumes one unit from a Mongo-backed fixed-window rate limit.
 *
 * Uses `findOneAndUpdate` with upsert and retries duplicate-key races once.
 * Infrastructure errors fail closed by default; set `failOpen: true` when
 * availability is more important than strict throttling for that flow.
 *
 * @param options - Counter options and Mongoose model.
 * @returns Rate-limit result.
 */
async function consumeMongoRateLimit(options: ConsumeMongoRateLimitOptions): Promise<RateLimitResult> {
  assertPositiveInteger(options.limit, 'limit');
  assertPositiveInteger(options.windowSeconds, 'windowSeconds');

  const nowMs = options.nowMs || Date.now();
  const windowStart = toRateLimitWindowStart(nowMs, options.windowSeconds);
  const resetAt = new Date(windowStart.getTime() + options.windowSeconds * 1000);
  const expireAt = new Date(windowStart.getTime() + options.windowSeconds * 1000 * (options.ttlWindowMultiplier || 2));
  const filter = {
    action: options.action,
    key: options.key,
    windowStart,
  };

  try {
    const entry = await options.model.findOneAndUpdate(
      filter,
      {
        $inc: { count: 1 },
        $setOnInsert: { expireAt },
      },
      {
        lean: true,
        new: true,
        setDefaultsOnInsert: true,
        upsert: true,
      }
    ) as unknown as MongoRateLimitDocument | null;

    return createResult(options.action, entry?.count || 1, options.limit, windowStart, resetAt, nowMs);
  } catch (error) {
    if (isDuplicateKeyError(error)) {
      const entry = await options.model.findOneAndUpdate(
        filter,
        { $inc: { count: 1 } },
        {
          lean: true,
          new: true,
        }
      ) as unknown as MongoRateLimitDocument | null;

      return createResult(options.action, entry?.count || 1, options.limit, windowStart, resetAt, nowMs);
    }

    if (options.failOpen) {
      return createResult(options.action, 0, options.limit, windowStart, resetAt, nowMs);
    }

    throw error;
  }
}

/**
 * Builds the storage key parts for a single policy scope.
 *
 * @param scope - Policy scope.
 * @param options - Source options carrying event, identifier and accountId.
 * @returns Ordered key parts (or null if the scope cannot be evaluated).
 */
function getScopeKeyParts(
  scope: RateLimitScope,
  options: EnforceMongoRateLimitOptions
): Array<string | number | null | undefined> | null {
  switch (scope) {
    case 'ip':
      return [getClientIp(options.event)];
    case 'identifier':
      // identifier may be null/undefined for unauthenticated public flows; the
      // shared key builder maps that to the literal `'anonymous'` part, which
      // would create one global bucket. For the layered limiter we treat a
      // missing identifier as "skip this lane" instead, so anonymous traffic
      // is governed by the ip / ip+identifier / global lanes only.
      if (options.identifier === undefined || options.identifier === null || options.identifier === '') {
        return null;
      }
      return [String(options.identifier)];
    case 'ip+identifier':
      return [getClientIp(options.event), options.identifier];
    case 'account':
      if (options.accountId === undefined || options.accountId === null || options.accountId === '') {
        return null;
      }
      return [`account:${String(options.accountId)}`];
    case 'global':
      return ['global'];
    default:
      return null;
  }
}

/**
 * Consumes one unit for a single policy and returns its result.
 *
 * The storage `action` is suffixed with the scope (e.g. `auth.login:ip`) so
 * different policy lanes never share a counter row.
 *
 * @param policy - Single policy lane.
 * @param options - Shared enforce options.
 * @returns Rate-limit result, or null when the lane is skipped.
 */
function enforceSinglePolicy(
  policy: RateLimitPolicy,
  options: EnforceMongoRateLimitOptions
): Promise<RateLimitResult> | null {
  const keyParts = getScopeKeyParts(policy.scope, options);
  if (keyParts === null) {
    return null;
  }

  return consumeMongoRateLimit({
    action: `${options.action}:${policy.scope}`,
    failOpen: options.failOpen,
    key: createRateLimitKey(keyParts, {
      hash: options.hashKey !== false,
      salt: options.keySalt,
    }),
    limit: policy.limit,
    model: options.model,
    nowMs: options.nowMs,
    ttlWindowMultiplier: options.ttlWindowMultiplier,
    windowSeconds: policy.windowSeconds,
  });
}

/**
 * Event-level Mongo rate limiter.
 *
 * In layered mode (`policies` provided), every lane is enforced sequentially
 * and the first non-allowed result is returned without checking remaining
 * lanes. In legacy mode the previous IP+identifier behaviour is preserved.
 *
 * @param options - Event, action, identifier/accountId, policies (or legacy
 *   limit/windowSeconds), and model.
 * @returns Rate-limit result.
 */
async function enforceMongoRateLimit(options: EnforceMongoRateLimitOptions): Promise<RateLimitResult> {
  if (options.policies && options.policies.length > 0) {
    let lastResult: RateLimitResult | null = null;

    for (const policy of options.policies) {
      const result = await enforceSinglePolicy(policy, options);
      if (result === null) {
        continue;
      }

      lastResult = result;
      if (!result.allowed) {
        return result;
      }
    }

    if (lastResult) {
      return lastResult;
    }

    // No lane applied (e.g. all skipped). Build a benign allowed result so
    // callers do not see a phantom rejection.
    const nowMs = options.nowMs || Date.now();
    const fallbackWindow = options.policies[0].windowSeconds;
    const windowStart = toRateLimitWindowStart(nowMs, fallbackWindow);
    const resetAt = new Date(windowStart.getTime() + fallbackWindow * 1000);
    return createResult(options.action, 0, options.policies[0].limit, windowStart, resetAt, nowMs);
  }

  // Legacy single-policy mode: IP + identifier composite key, original action.
  if (options.limit === undefined || options.windowSeconds === undefined) {
    throw new Error('requireMongoRateLimit requires either `policies` or both `limit` and `windowSeconds`');
  }

  const keyParts = [
    options.includeIp === false ? undefined : getClientIp(options.event),
    options.identifier,
  ];

  return consumeMongoRateLimit({
    action: options.action,
    failOpen: options.failOpen,
    key: createRateLimitKey(keyParts, {
      hash: options.hashKey !== false,
      salt: options.keySalt,
    }),
    limit: options.limit,
    model: options.model,
    nowMs: options.nowMs,
    ttlWindowMultiplier: options.ttlWindowMultiplier,
    windowSeconds: options.windowSeconds,
  });
}

/**
 * Requires the current request to be within rate limit.
 *
 * Supports two forms:
 *
 * 1. **Layered** (recommended): pass `policies: RateLimitPolicy[]`. Each lane
 *    is checked independently and the request is rejected on the first lane
 *    that is over its limit. Use this to combine per-IP, per-identifier and
 *    per-account caps so distributed-IP attacks against a single victim are
 *    bounded.
 * 2. **Legacy shorthand**: pass `limit` and `windowSeconds`. Behaviour is
 *    identical to the original single IP+identifier composite limiter.
 *
 * Throws `RateLimitError` with `statusCode: 429` and
 * `errorKey: common.rateLimited` when exceeded.
 *
 * @param options - Event-level Mongo rate-limit options.
 */
export async function requireMongoRateLimit(options: RequireMongoRateLimitOptions): Promise<void> {
  const result = await enforceMongoRateLimit({
    ...options,
    model: getMongoRateLimitModel(options.mongoose, {
      collectionName: options.collectionName,
      modelName: options.modelName,
    }),
  });

  if (!result.allowed) {
    throw new RateLimitError(result);
  }
}

/**
 * Peeks at a failure counter without incrementing it and throws
 * `RateLimitError` if the cooldown threshold has already been reached for the
 * current window.
 *
 * Use at the top of a sensitive handler so legitimate requests do not consume
 * the failure quota; pair with {@link recordMongoRateLimitFailure} on actual
 * authentication failures.
 *
 * @param options - Failure counter options.
 */
export async function requireMongoRateLimitNotInCooldown(options: FailureCounterOptions): Promise<void> {
  const scope = options.scope || 'identifier';
  const keyParts = getScopeKeyParts(scope, {
    accountId: options.accountId,
    action: options.action,
    event: options.event,
    failOpen: options.failOpen,
    hashKey: options.hashKey,
    identifier: options.identifier,
    keySalt: options.keySalt,
    model: getMongoRateLimitModel(options.mongoose, {
      collectionName: options.collectionName,
      modelName: options.modelName,
    }),
    nowMs: options.nowMs,
    ttlWindowMultiplier: options.ttlWindowMultiplier,
  });

  // Skip if the key cannot be built (e.g. anonymous identifier scope).
  if (keyParts === null) {
    return;
  }

  const model = getMongoRateLimitModel(options.mongoose, {
    collectionName: options.collectionName,
    modelName: options.modelName,
  });
  const nowMs = options.nowMs || Date.now();
  const windowStart = toRateLimitWindowStart(nowMs, options.cooldownSeconds);
  const resetAt = new Date(windowStart.getTime() + options.cooldownSeconds * 1000);
  const key = createRateLimitKey(keyParts, {
    hash: options.hashKey !== false,
    salt: options.keySalt,
  });

  let entry: MongoRateLimitDocument | null = null;
  try {
    entry = await model.findOne({
      action: `${options.action}:${scope}`,
      key,
      windowStart,
    }).lean() as unknown as MongoRateLimitDocument | null;
  } catch (error) {
    if (options.failOpen) {
      return;
    }
    throw error;
  }

  const count = entry?.count || 0;
  const result = createResult(`${options.action}:${scope}`, count, options.threshold, windowStart, resetAt, nowMs);
  if (count >= options.threshold) {
    throw new RateLimitError(result);
  }
}

/**
 * Records one failure for the configured cooldown counter and returns the
 * current state. The caller decides what counts as a failure (e.g. wrong
 * password, invalid OTP, bad refresh token). Successful requests must not
 * call this so they do not consume the failure quota.
 *
 * Implementation note: stored as a regular fixed-window counter where the
 * "limit" is the cooldown threshold and the "window" is the cooldown length.
 * Once `count >= threshold` within the window, subsequent
 * {@link requireMongoRateLimitNotInCooldown} calls will reject the request.
 *
 * @param options - Failure counter options.
 * @returns Latest count, whether the cooldown was triggered, and retry-after.
 */
export async function recordMongoRateLimitFailure(
  options: FailureCounterOptions
): Promise<{ cooldownTriggered: boolean; count: number; retryAfterSeconds: number }> {
  const scope = options.scope || 'identifier';
  const model = getMongoRateLimitModel(options.mongoose, {
    collectionName: options.collectionName,
    modelName: options.modelName,
  });

  const keyParts = getScopeKeyParts(scope, {
    accountId: options.accountId,
    action: options.action,
    event: options.event,
    failOpen: options.failOpen,
    hashKey: options.hashKey,
    identifier: options.identifier,
    keySalt: options.keySalt,
    model,
    nowMs: options.nowMs,
    ttlWindowMultiplier: options.ttlWindowMultiplier,
  });

  if (keyParts === null) {
    return { cooldownTriggered: false, count: 0, retryAfterSeconds: 0 };
  }

  const result = await consumeMongoRateLimit({
    action: `${options.action}:${scope}`,
    failOpen: options.failOpen,
    key: createRateLimitKey(keyParts, {
      hash: options.hashKey !== false,
      salt: options.keySalt,
    }),
    limit: options.threshold,
    model,
    nowMs: options.nowMs,
    ttlWindowMultiplier: options.ttlWindowMultiplier,
    windowSeconds: options.cooldownSeconds,
  });

  return {
    cooldownTriggered: result.count >= options.threshold,
    count: result.count,
    retryAfterSeconds: result.retryAfterSeconds,
  };
}

export type { RateLimitPolicy, RateLimitScope };
