// @aws-ddd-api/shared/rate-limit/mongo.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';
import type { Model, Mongoose, Schema } from 'mongoose';

declare const require: (moduleName: string) => {
  createHash: (algorithm: string) => {
    digest: (encoding: 'hex') => string;
    update: (data: string) => ReturnType<ReturnType<typeof require>['createHash']>;
  };
};

const { createHash } = require('crypto');

/**
 * MongoDB rate-limit document shape.
 */
interface MongoRateLimitDocument {
  action: string;
  key: string;
  windowStart: Date;
  count: number;
  expireAt: Date;
}

/**
 * Mongoose model used by the shared Mongo rate limiter.
 */
type MongoRateLimitModel = Model<MongoRateLimitDocument>;

/**
 * Internal options for creating or retrieving the shared rate-limit model.
 */
interface GetMongoRateLimitModelOptions {
  collectionName?: string;
  modelName?: string;
}

/**
 * Internal options for hashing rate-limit key material before storing it.
 */
interface HashRateLimitKeyOptions {
  salt?: string;
}

/**
 * Internal options for building a normalized rate-limit key.
 */
interface CreateRateLimitKeyOptions extends HashRateLimitKeyOptions {
  hash?: boolean;
  separator?: string;
}

/**
 * Result returned after consuming one rate-limit unit.
 */
interface RateLimitResult {
  action: string;
  allowed: boolean;
  count: number;
  limit: number;
  remaining: number;
  resetAt: string;
  retryAfterSeconds: number;
  windowStart: string;
}

/**
 * Internal options for the low-level Mongo counter operation.
 */
interface ConsumeMongoRateLimitOptions {
  action: string;
  failOpen?: boolean;
  key: string;
  limit: number;
  model: MongoRateLimitModel;
  nowMs?: number;
  ttlWindowMultiplier?: number;
  windowSeconds: number;
}

/**
 * Options for event-based rate limiting.
 */
interface EnforceMongoRateLimitOptions {
  action: string;
  event: APIGatewayProxyEvent;
  failOpen?: boolean;
  hashKey?: boolean;
  identifier?: string | number | null;
  includeIp?: boolean;
  keySalt?: string;
  limit: number;
  model: MongoRateLimitModel;
  nowMs?: number;
  ttlWindowMultiplier?: number;
  windowSeconds: number;
}

/**
 * Options for requiring a Mongo-backed rate limit for the current request.
 */
interface RequireMongoRateLimitOptions extends Omit<EnforceMongoRateLimitOptions, 'model'> {
  collectionName?: string;
  modelName?: string;
  mongoose: Mongoose;
}

/**
 * Internal error thrown by `requireMongoRateLimit` when a request is over limit.
 */
class RateLimitError extends Error {
  errorKey = 'common.rateLimited';
  result: RateLimitResult;
  statusCode = 429;

  constructor(result: RateLimitResult) {
    super('common.rateLimited');
    this.name = 'RateLimitError';
    this.result = result;
  }
}

/**
 * Reads a header from an API Gateway event without depending on header casing.
 *
 * @param event - API Gateway event.
 * @param name - Header name to read.
 * @returns Header value, or undefined when not present.
 */
function getHeaderValue(event: APIGatewayProxyEvent, name: string): string | undefined {
  const target = name.toLowerCase();
  const headers = event.headers || {};

  for (const [key, value] of Object.entries(headers)) {
    if (key.toLowerCase() === target && typeof value === 'string') {
      return value;
    }
  }

  return undefined;
}

/**
 * Extracts the client IP from an API Gateway event.
 *
 * Prefers the first `X-Forwarded-For` value and falls back to REST/HTTP API
 * request context source IPs.
 *
 * @param event - API Gateway event.
 * @returns Client IP, or `unknown` when not available.
 */
function getClientIp(event: APIGatewayProxyEvent): string {
  const forwardedFor = getHeaderValue(event, 'x-forwarded-for');
  if (typeof forwardedFor === 'string' && forwardedFor.trim()) {
    return forwardedFor.split(',')[0].trim();
  }

  return event.requestContext?.identity?.sourceIp || 'unknown';
}

/**
 * Floors a timestamp to the start of its fixed rate-limit window.
 *
 * @param nowMs - Current epoch milliseconds.
 * @param windowSeconds - Window length in seconds.
 * @returns Window start date.
 */
function toRateLimitWindowStart(nowMs: number, windowSeconds: number): Date {
  const windowMs = windowSeconds * 1000;
  return new Date(Math.floor(nowMs / windowMs) * windowMs);
}

/**
 * Hashes rate-limit key material for safer persistence.
 *
 * Use this for email, phone, token, IP, or other user-identifying key parts.
 *
 * @param value - Raw key material.
 * @param options - Optional salt.
 * @returns SHA-256 hex digest.
 */
function hashRateLimitKey(value: string, options: HashRateLimitKeyOptions = {}): string {
  return createHash('sha256')
    .update(options.salt || '')
    .update(value)
    .digest('hex');
}

/**
 * Builds a deterministic rate-limit key from multiple key parts.
 *
 * By default the returned key is hashed so raw identifiers are not stored.
 *
 * @param parts - Key parts such as client IP, user id, email, or route name.
 * @param options - Hashing and separator options.
 * @returns Stable key for Mongo rate-limit storage.
 */
function createRateLimitKey(
  parts: Array<string | number | null | undefined>,
  options: CreateRateLimitKeyOptions = {}
): string {
  const separator = options.separator || ':';
  const rawKey = parts
    .map((part) => (part === undefined || part === null || part === '' ? 'anonymous' : String(part).trim()))
    .join(separator);

  return options.hash === false ? rawKey : hashRateLimitKey(rawKey, options);
}

/**
 * Creates the Mongo rate-limit schema.
 *
 * The schema uses a unique fixed-window index and a TTL index on `expireAt`.
 *
 * @param mongoose - Domain Lambda's connected/imported Mongoose instance.
 * @returns Mongoose schema for rate-limit counters.
 */
function createMongoRateLimitSchema(mongoose: Mongoose): Schema<MongoRateLimitDocument> {
  const schema = new mongoose.Schema<MongoRateLimitDocument>(
    {
      action: {
        required: true,
        trim: true,
        type: String,
      },
      key: {
        required: true,
        trim: true,
        type: String,
      },
      windowStart: {
        required: true,
        type: Date,
      },
      count: {
        default: 0,
        required: true,
        type: Number,
      },
      expireAt: {
        required: true,
        type: Date,
      },
    },
    {
      timestamps: true,
    }
  );

  schema.index({ action: 1, key: 1, windowStart: 1 }, { unique: true });
  schema.index({ expireAt: 1 }, { expireAfterSeconds: 0 });

  return schema;
}

/**
 * Gets or registers the Mongo rate-limit model on a domain Mongoose instance.
 *
 * Call this after the Lambda's normal DB setup has imported/connected
 * Mongoose.
 *
 * @param mongoose - Domain Lambda's Mongoose instance.
 * @param options - Optional model and collection names.
 * @returns Mongoose model for rate-limit counters.
 */
function getMongoRateLimitModel(
  mongoose: Mongoose,
  options: GetMongoRateLimitModelOptions = {}
): MongoRateLimitModel {
  const modelName = options.modelName || 'RateLimit';
  const collectionName = options.collectionName || 'rate_limits';
  const existingModel = mongoose.models[modelName] as MongoRateLimitModel | undefined;

  return existingModel || mongoose.model<MongoRateLimitDocument>(
    modelName,
    createMongoRateLimitSchema(mongoose),
    collectionName
  );
}

function getRetryAfterSeconds(nowMs: number, resetAt: Date): number {
  return Math.max(0, Math.ceil((resetAt.getTime() - nowMs) / 1000));
}

function createResult(
  action: string,
  count: number,
  limit: number,
  windowStart: Date,
  resetAt: Date,
  nowMs: number
): RateLimitResult {
  return {
    action,
    allowed: count <= limit,
    count,
    limit,
    remaining: Math.max(0, limit - count),
    resetAt: resetAt.toISOString(),
    retryAfterSeconds: getRetryAfterSeconds(nowMs, resetAt),
    windowStart: windowStart.toISOString(),
  };
}

function isDuplicateKeyError(error: unknown): boolean {
  return (error as { code?: unknown })?.code === 11000;
}

function assertPositiveInteger(value: number, name: string): void {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer`);
  }
}

/**
 * Atomically consumes one unit from a Mongo-backed fixed-window rate limit.
 *
 * Uses `findOneAndUpdate` with upsert and retries duplicate-key races once.
 * Infrastructure errors fail closed by default; set `failOpen: true` when
 * availability is more important than strict throttling for that flow.
 *
 * @param options - Counter options and Mongoose model.
 * @returns Rate-limit result.
 */
async function consumeMongoRateLimit(options: ConsumeMongoRateLimitOptions): Promise<RateLimitResult> {
  assertPositiveInteger(options.limit, 'limit');
  assertPositiveInteger(options.windowSeconds, 'windowSeconds');

  const nowMs = options.nowMs || Date.now();
  const windowStart = toRateLimitWindowStart(nowMs, options.windowSeconds);
  const resetAt = new Date(windowStart.getTime() + options.windowSeconds * 1000);
  const expireAt = new Date(windowStart.getTime() + options.windowSeconds * 1000 * (options.ttlWindowMultiplier || 2));
  const filter = {
    action: options.action,
    key: options.key,
    windowStart,
  };

  try {
    const entry = await options.model.findOneAndUpdate(
      filter,
      {
        $inc: { count: 1 },
        $setOnInsert: { expireAt },
      },
      {
        lean: true,
        new: true,
        setDefaultsOnInsert: true,
        upsert: true,
      }
    ) as unknown as MongoRateLimitDocument | null;

    return createResult(options.action, entry?.count || 1, options.limit, windowStart, resetAt, nowMs);
  } catch (error) {
    if (isDuplicateKeyError(error)) {
      const entry = await options.model.findOneAndUpdate(
        filter,
        { $inc: { count: 1 } },
        {
          lean: true,
          new: true,
        }
      ) as unknown as MongoRateLimitDocument | null;

      return createResult(options.action, entry?.count || 1, options.limit, windowStart, resetAt, nowMs);
    }

    if (options.failOpen) {
      return createResult(options.action, 0, options.limit, windowStart, resetAt, nowMs);
    }

    throw error;
  }
}

/**
 * Event-level Mongo rate limiter using client IP plus optional identifier.
 *
 * @param options - Event, action, identifier, limit, and model.
 * @returns Rate-limit result.
 */
function enforceMongoRateLimit(options: EnforceMongoRateLimitOptions): Promise<RateLimitResult> {
  const keyParts = [
    options.includeIp === false ? undefined : getClientIp(options.event),
    options.identifier,
  ];

  return consumeMongoRateLimit({
    ...options,
    key: createRateLimitKey(keyParts, {
      hash: options.hashKey !== false,
      salt: options.keySalt,
    }),
  });
}

/**
 * Requires the current request to be within rate limit.
 *
 * Throws `RateLimitError` with `statusCode: 429` and
 * `errorKey: common.rateLimited` when exceeded, so the shared handler formats a
 * standardized localized response.
 *
 * @param options - Event-level Mongo rate-limit options.
 */
export async function requireMongoRateLimit(options: RequireMongoRateLimitOptions): Promise<void> {
  const result = await enforceMongoRateLimit({
    ...options,
    model: getMongoRateLimitModel(options.mongoose, {
      collectionName: options.collectionName,
      modelName: options.modelName,
    }),
  });

  if (!result.allowed) {
    throw new RateLimitError(result);
  }
}
