// @aws-ddd-api/shared/index.js
'use strict';

function json(statusCode, payload, extraHeaders = {}) {
  return {
    statusCode,
    headers: {
      'content-type': 'application/json',
      ...extraHeaders,
    },
    body: JSON.stringify(payload),
  };
}

function safeJsonParse(body) {
  if (!body) {
    return null;
  }

  try {
    return JSON.parse(body);
  } catch {
    return body;
  }
}

function isTrue(value, defaultValue = false) {
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }

  const normalized = String(value).trim().toLowerCase();
  return normalized === '1' || normalized === 'true' || normalized === 'yes' || normalized === 'on';
}

function getBearerToken(headerValue) {
  if (!headerValue) {
    return null;
  }

  const raw = String(headerValue).trim();
  const match = /^Bearer\s+(.+)$/.exec(raw);
  return match ? match[1] : null;
}

function buildPolicy({ principalId, effect, resource, context = {} }) {
  return {
    principalId,
    policyDocument: {
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: resource,
        },
      ],
    },
    context: Object.fromEntries(
      Object.entries(context).map(([key, value]) => [key, value == null ? '' : String(value)])
    ),
  };
}

module.exports = {
  buildPolicy,
  getBearerToken,
  isTrue,
  json,
  safeJsonParse,
};
