// @aws-ddd-api/shared/index.ts
export { getBearerToken } from './auth/bearer';
export type { AuthContext } from './auth/context';
export { AuthContextError, getAuthContext, hasRole, isSelf, requireAuthContext, requireRole } from './auth/context';
export type { PolicyInput } from './auth/policy';
export { buildPolicy } from './auth/policy';
export { isTrue } from './config/boolean';
export type { ApiRouteContext, ApiRouteRequest, CreateApiGatewayHandlerOptions, JsonResponder } from './http/handler';
export { createApiGatewayHandler } from './http/handler';
export type { CreateRouterOptions, RouteHandler, RouteMap } from './http/router';
export { createRouter } from './http/router';
export type { CorsHeaderOptions, ErrorJsonOptions, JsonResponse } from './http/response';
export { errorJson, json, safeJsonParse, withCorsHeaders } from './http/response';
export type { LogOptions } from './logging/logger';
export { logError, logInfo, logWarn, serializeError } from './logging/logger';
export {
  getFirstZodIssueMessage,
  getJoinedZodIssueMessages,
  getZodIssues,
  parseJsonBodyWithSchema,
} from './validation/zod';
