// @aws-ddd-api/shared/index.ts
export interface JsonResponse {
  statusCode: number;
  headers: Record<string, string>;
  body: string;
}

export function json(statusCode: number, payload: unknown, extraHeaders: Record<string, string> = {}): JsonResponse {
  return {
    statusCode,
    headers: {
      'content-type': 'application/json',
      ...extraHeaders,
    },
    body: JSON.stringify(payload),
  };
}

export function safeJsonParse(body: string | null | undefined): unknown {
  if (!body) {
    return null;
  }

  try {
    return JSON.parse(body);
  } catch {
    return body;
  }
}

export function isTrue(value: unknown, defaultValue = false): boolean {
  if (value === undefined || value === null || value === '') {
    return defaultValue;
  }

  const normalized = String(value).trim().toLowerCase();
  return normalized === '1' || normalized === 'true' || normalized === 'yes' || normalized === 'on';
}

export function getBearerToken(headerValue: unknown): string | null {
  if (!headerValue) {
    return null;
  }

  const raw = String(headerValue).trim();
  const match = /^Bearer\s+(.+)$/.exec(raw);
  return match ? match[1] : null;
}

export interface PolicyInput {
  principalId: string;
  effect: 'Allow' | 'Deny';
  resource: string;
  context?: Record<string, unknown>;
}

export function buildPolicy({ principalId, effect, resource, context = {} }: PolicyInput) {
  return {
    principalId,
    policyDocument: {
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: resource,
        },
      ],
    },
    context: Object.fromEntries(
      Object.entries(context).map(([key, value]) => [key, value == null ? '' : String(value)])
    ),
  };
}
