// @aws-ddd-api/shared/sanitization/text.ts
/**
 * Lightweight text sanitization helpers for free-text fields persisted to
 * MongoDB and later rendered by frontends.
 *
 * These helpers do NOT replace per-render escaping (frontends should still
 * escape when injecting into HTML/JSX). They harden storage by:
 *
 * - removing HTML/script tags before persistence (prevents XSS-at-rest in
 *   clients that mistakenly render raw values),
 * - stripping ASCII/Unicode control characters that have no legitimate place
 *   in user input,
 * - normalizing Unicode (NFC) so visually-identical strings compare equal.
 *
 * Sanitization is intentionally text-only. We do NOT preserve any HTML tags
 * because the API has no rich-text storage requirement.
 */

const TAG_REGEX = /<[^>]*>/g;
// Strip C0 controls (except \t, \n, \r) and DEL, plus C1 controls.
const CONTROL_CHAR_REGEX = /[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F-\u009F]/g;
const HTML_ENTITY_MAP: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
  '/': '&#x2F;',
};

/**
 * Escapes the six HTML-significant characters. Use only when echoing
 * user-supplied text directly into HTML on the server. Stored fields should
 * use {@link sanitizeText} instead and rely on frontend escaping for display.
 */
export function escapeHtml(input: string): string {
  return input.replace(/[&<>"'/]/g, (ch) => HTML_ENTITY_MAP[ch] ?? ch);
}

/**
 * Sanitizes free-text input destined for persistence:
 *
 * 1. Strip HTML/script tags entirely (no allow-list).
 * 2. Strip ASCII/Unicode control characters.
 * 3. Normalize to Unicode NFC.
 * 4. Trim leading/trailing whitespace.
 *
 * Returns an empty string for `null`/`undefined`/non-string input.
 */
export function sanitizeText(input: unknown): string {
  if (typeof input !== 'string') return '';
  return input
    .replace(TAG_REGEX, '')
    .replace(CONTROL_CHAR_REGEX, '')
    .normalize('NFC')
    .trim();
}

/**
 * Same as {@link sanitizeText} but preserves `null`/`undefined`. Useful for
 * PATCH bodies where omitting a field and clearing it are distinct operations.
 */
export function sanitizeOptionalText<T extends string | null | undefined>(
  input: T
): T extends string ? string : T {
  if (input === null || input === undefined) return input as never;
  return sanitizeText(input) as never;
}

/**
 * Applies {@link sanitizeOptionalText} to the listed string fields of an
 * object in-place and returns it. Non-string values are left untouched.
 *
 * ```ts
 * sanitizeFields(data, ['description', 'remarks']);
 * ```
 */
export function sanitizeFields<T extends Record<string, unknown>>(
  obj: T,
  fields: readonly (keyof T)[]
): T {
  for (const field of fields) {
    const value = obj[field];
    if (typeof value === 'string') {
      (obj as Record<string, unknown>)[field as string] = sanitizeText(value);
    }
  }
  return obj;
}
