// @aws-ddd-api/shared/auth/context.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';
import jwt from 'jsonwebtoken';
import { getBearerToken } from './bearer';
import { isTrue } from '../config/boolean';

/**
 * Normalized identity claims produced by the API Gateway Lambda authorizer.
 */
export interface AuthContext {
  authMode?: string;
  stage?: string;
  userId: string;
  userEmail?: string;
  userRole?: string;
  ngoId?: string;
  ngoName?: string;
  principalId?: string;
}

/**
 * Error type thrown by auth context guards.
 *
 * `createApiGatewayHandler` reads `statusCode`, so throwing this from domain
 * code maps naturally to a 401 or 403 response.
 */
export class AuthContextError extends Error {
  statusCode: number;
  errorKey: string;

  constructor(errorKey: string, statusCode: number) {
    super(errorKey);
    this.name = 'AuthContextError';
    this.errorKey = errorKey;
    this.statusCode = statusCode;
  }
}

/**
 * Reads a single authorizer context claim as a string.
 *
 * @param authorizer - Raw API Gateway authorizer context object.
 * @param key - Claim name to read.
 * @returns A string claim value, or `undefined` when missing or empty.
 */
function getStringClaim(authorizer: Record<string, unknown>, key: string): string | undefined {
  const value = authorizer[key];
  if (value === undefined || value === null || value === '') {
    return undefined;
  }

  return String(value);
}

/**
 * Returns the raw API Gateway authorizer context from a REST API event.
 *
 * @param event - API Gateway proxy event.
 * @returns Raw authorizer context or `null` when the route was unauthenticated.
 */
function getAuthorizerContext(event: APIGatewayProxyEvent): Record<string, unknown> | null {
  const authorizer = event.requestContext?.authorizer;
  if (!authorizer || typeof authorizer !== 'object') {
    if (isTrue(process.env.AWS_SAM_LOCAL, false)) {
      const token = getBearerToken(
        event.headers?.Authorization || event.headers?.authorization
      );
      const jwtSecret = process.env.JWT_SECRET || '';

      if (token && jwtSecret) {
        try {
          const decoded = jwt.verify(token, jwtSecret, { algorithms: ['HS256'] }) as Record<
            string,
            unknown
          >;
          const userId = decoded.userId || decoded.sub;

          if (userId) {
            return {
              authMode: 'sam-local-jwt-fallback',
              stage: process.env.STAGE_NAME || 'development',
              userId: String(userId),
              userEmail:
                decoded.userEmail || decoded.email ? String(decoded.userEmail || decoded.email) : '',
              userRole:
                decoded.userRole || decoded.role ? String(decoded.userRole || decoded.role) : '',
              ngoId: decoded.ngoId ? String(decoded.ngoId) : '',
              ngoName: decoded.ngoName ? String(decoded.ngoName) : '',
              principalId: String(userId),
            };
          }
        } catch {
          return null;
        }
      }

      if (isTrue(process.env.AUTH_BYPASS, false)) {
        return {
          authMode: 'bypass',
          stage: process.env.STAGE_NAME || 'development',
          userId: '000000000000000000000000',
          userEmail: 'dev@test.com',
          userRole: 'developer',
          principalId: '000000000000000000000000',
        };
      }
    }

    return null;
  }

  return authorizer as Record<string, unknown>;
}

/**
 * Extracts normalized auth claims from `event.requestContext.authorizer`.
 *
 * @param event - API Gateway proxy event after the Lambda authorizer has run.
 * @returns Normalized auth context, or `null` when no user id is available.
 */
export function getAuthContext(event: APIGatewayProxyEvent): AuthContext | null {
  const authorizer = getAuthorizerContext(event);
  if (!authorizer) {
    return null;
  }

  const userId = getStringClaim(authorizer, 'userId') || getStringClaim(authorizer, 'principalId');
  if (!userId) {
    return null;
  }

  return {
    authMode: getStringClaim(authorizer, 'authMode'),
    stage: getStringClaim(authorizer, 'stage'),
    userId,
    userEmail: getStringClaim(authorizer, 'userEmail'),
    userRole: getStringClaim(authorizer, 'userRole'),
    ngoId: getStringClaim(authorizer, 'ngoId'),
    ngoName: getStringClaim(authorizer, 'ngoName'),
    principalId: getStringClaim(authorizer, 'principalId'),
  };
}

/**
 * Requires a valid authorizer context and returns it.
 *
 * @param event - API Gateway proxy event.
 * @returns Normalized auth context.
 * @throws AuthContextError with status 401 when no authenticated user context exists.
 */
export function requireAuthContext(event: APIGatewayProxyEvent): AuthContext {
  const authContext = getAuthContext(event);
  if (!authContext) {
    throw new AuthContextError('common.unauthorized', 401);
  }

  return authContext;
}

/**
 * Checks whether the authenticated user has one exact role.
 *
 * @param event - API Gateway proxy event.
 * @param role - Required role name.
 * @returns True when `event.requestContext.authorizer.userRole` matches `role`.
 */
export function hasRole(event: APIGatewayProxyEvent, role: string): boolean {
  const authContext = getAuthContext(event);
  return authContext?.userRole === role;
}

/**
 * Requires the authenticated user to have one of the allowed roles.
 *
 * @param event - API Gateway proxy event.
 * @param roles - One role or a list of accepted roles.
 * @returns Normalized auth context when the role is allowed.
 * @throws AuthContextError with status 401 when unauthenticated, or 403 when the role is not allowed.
 */
export function requireRole(event: APIGatewayProxyEvent, roles: string | string[]): AuthContext {
  const authContext = requireAuthContext(event);
  const allowedRoles = Array.isArray(roles) ? roles : [roles];

  if (!authContext.userRole || !allowedRoles.includes(authContext.userRole)) {
    throw new AuthContextError('common.forbidden', 403);
  }

  return authContext;
}
