// @aws-ddd-api/shared/auth/context.ts
import type { APIGatewayProxyEvent } from 'aws-lambda';

export interface AuthContext {
  authMode?: string;
  stage?: string;
  userId: string;
  userEmail?: string;
  userRole?: string;
  ngoId?: string;
  ngoName?: string;
  principalId?: string;
}

export class AuthContextError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = 'AuthContextError';
    this.statusCode = statusCode;
  }
}

function getStringClaim(authorizer: Record<string, unknown>, key: string): string | undefined {
  const value = authorizer[key];
  if (value === undefined || value === null || value === '') {
    return undefined;
  }

  return String(value);
}

function getAuthorizerContext(event: APIGatewayProxyEvent): Record<string, unknown> | null {
  const authorizer = event.requestContext?.authorizer;
  if (!authorizer || typeof authorizer !== 'object') {
    return null;
  }

  return authorizer as Record<string, unknown>;
}

export function getAuthContext(event: APIGatewayProxyEvent): AuthContext | null {
  const authorizer = getAuthorizerContext(event);
  if (!authorizer) {
    return null;
  }

  const userId = getStringClaim(authorizer, 'userId') || getStringClaim(authorizer, 'principalId');
  if (!userId) {
    return null;
  }

  return {
    authMode: getStringClaim(authorizer, 'authMode'),
    stage: getStringClaim(authorizer, 'stage'),
    userId,
    userEmail: getStringClaim(authorizer, 'userEmail'),
    userRole: getStringClaim(authorizer, 'userRole'),
    ngoId: getStringClaim(authorizer, 'ngoId'),
    ngoName: getStringClaim(authorizer, 'ngoName'),
    principalId: getStringClaim(authorizer, 'principalId'),
  };
}

export function requireAuthContext(event: APIGatewayProxyEvent): AuthContext {
  const authContext = getAuthContext(event);
  if (!authContext) {
    throw new AuthContextError('Unauthorized', 401);
  }

  return authContext;
}

export function hasRole(event: APIGatewayProxyEvent, role: string): boolean {
  const authContext = getAuthContext(event);
  return authContext?.userRole === role;
}

export function requireRole(event: APIGatewayProxyEvent, roles: string | string[]): AuthContext {
  const authContext = requireAuthContext(event);
  const allowedRoles = Array.isArray(roles) ? roles : [roles];

  if (!authContext.userRole || !allowedRoles.includes(authContext.userRole)) {
    throw new AuthContextError('Forbidden', 403);
  }

  return authContext;
}

export function isSelf(event: APIGatewayProxyEvent, userId: string | number | null | undefined): boolean {
  const authContext = getAuthContext(event);
  return Boolean(authContext && userId != null && String(authContext.userId) === String(userId));
}
