// @aws-ddd-api/shared/auth/policy.ts
/**
 * Input used to build an API Gateway Lambda authorizer IAM policy response.
 */
export interface PolicyInput {
  principalId: string;
  effect: 'Allow' | 'Deny';
  resource: string;
  context?: Record<string, unknown>;
}

/**
 * Builds the policy object expected by REST API Gateway Lambda authorizers.
 *
 * API Gateway authorizer context values must be strings, so `context` values are
 * converted to strings and nullish values become empty strings.
 *
 * @param input - Principal, allow/deny effect, target method ARN, and optional authorizer context.
 * @returns An API Gateway authorizer policy response.
 */
export function buildPolicy({ principalId, effect, resource, context = {} }: PolicyInput) {
  return {
    principalId,
    policyDocument: {
      Version: '2012-10-17',
      Statement: [
        {
          Action: 'execute-api:Invoke',
          Effect: effect,
          Resource: resource,
        },
      ],
    },
    context: Object.fromEntries(
      Object.entries(context).map(([key, value]) => [key, value == null ? '' : String(value)])
    ),
  };
}
