// @aws-ddd-api/shared/validation/common.ts
import { z } from 'zod';
import mongoose from 'mongoose';

/**
 * Common Zod validators reused across domains.
 *
 * Centralizing these prevents drift in formats, length limits, and error keys.
 * All validators emit dotted i18n error keys so `parseBody` can surface them
 * directly through `errorResponse`.
 */

// -- Identifiers ------------------------------------------------------------

/**
 * 24-char hex MongoDB ObjectId. Use for path params and body fields that are
 * persisted as `_id` references. Returns the string unchanged on success.
 */
export const objectIdString = (errorKey: string = 'common.invalidObjectId') =>
  z
    .string({ error: errorKey })
    .trim()
    .refine((value) => mongoose.isValidObjectId(value), { message: errorKey });

/**
 * Short opaque identifier (alphanumeric + `_-`) up to 64 chars. Suitable for
 * order/temporary IDs that are generated client-side or by external systems.
 */
export const tempIdString = (errorKey: string = 'common.invalidPathParams') =>
  z
    .string({ error: errorKey })
    .trim()
    .min(1, errorKey)
    .max(64, errorKey)
    .regex(/^[A-Za-z0-9_-]+$/, errorKey);

// -- Common formats ---------------------------------------------------------

/** RFC-5322-ish email (Zod's built-in validator), lowercased + trimmed. */
export const emailString = (errorKey: string = 'common.invalidEmail') =>
  z
    .string({ error: errorKey })
    .trim()
    .toLowerCase()
    .max(254, errorKey)
    .email(errorKey);

/** E.164 phone number (`+` followed by 1-15 digits, leading digit non-zero). */
export const phoneE164String = (errorKey: string = 'common.invalidPhone') =>
  z
    .string({ error: errorKey })
    .trim()
    .regex(/^\+[1-9]\d{1,14}$/, errorKey);

/** Absolute http(s) URL up to 2048 chars. */
export const urlString = (errorKey: string = 'common.invalidUrl') =>
  z
    .string({ error: errorKey })
    .trim()
    .max(2048, errorKey)
    .url(errorKey)
    .refine((v) => /^https?:\/\//i.test(v), { message: errorKey });

// -- Pagination -------------------------------------------------------------

export interface PaginationQueryOptions {
  /** Maximum allowed `limit`. Default: 100. */
  maxLimit?: number;
  /** Default `limit` when not provided. Default: 30. */
  defaultLimit?: number;
  /** Maximum allowed `page`. Default: 10000 (prevents pagination DoS). */
  maxPage?: number;
  /** i18n error key for invalid pagination params. */
  errorKey?: string;
}

/**
 * Returns a Zod schema for `{ page, limit }` query strings.
 *
 * - Coerces strings (`"1"`) to numbers
 * - Enforces integer + positive bounds
 * - Caps `page` at `maxPage` and `limit` at `maxLimit` to prevent DoS
 * - Falls back to `defaultLimit` and `page=1` when the key is missing
 *
 * Apply against `event.queryStringParameters ?? {}` — the resulting object is
 * always defined, even when the caller passes nothing.
 */
export const paginationQuerySchema = (options: PaginationQueryOptions = {}) => {
  const {
    maxLimit = 100,
    defaultLimit = 30,
    maxPage = 10000,
    errorKey = 'common.invalidQueryParams',
  } = options;
  return z
    .object({
      page: z.coerce
        .number({ error: errorKey })
        .int(errorKey)
        .min(1, errorKey)
        .max(maxPage, errorKey)
        .optional()
        .default(1),
      limit: z.coerce
        .number({ error: errorKey })
        .int(errorKey)
        .min(1, errorKey)
        .max(maxLimit, errorKey)
        .optional()
        .default(defaultLimit),
    })
    .passthrough();
};

// -- Path param helper ------------------------------------------------------

export interface PathParamFailure {
  ok: false;
  statusCode: number;
  errorKey: string;
}
export interface PathParamSuccess<T> {
  ok: true;
  data: T;
}
export type PathParamResult<T> = PathParamSuccess<T> | PathParamFailure;

/**
 * Validates a single path parameter against a Zod schema.
 *
 * Returns a discriminated union so callers can short-circuit with the standard
 * 400 + i18n error key pattern, mirroring `parseBody`:
 *
 * ```ts
 * const r = parsePathParam(ctx.event.pathParameters?.petId, objectIdString());
 * if (!r.ok) return response.errorResponse(r.statusCode, r.errorKey, ctx.event);
 * const petId = r.data;
 * ```
 */
export function parsePathParam<T>(
  raw: unknown,
  schema: z.ZodType<T>,
  fallbackErrorKey: string = 'common.invalidPathParams'
): PathParamResult<T> {
  const parsed = schema.safeParse(raw);
  if (!parsed.success) {
    const first = parsed.error.issues[0];
    const message = typeof first?.message === 'string' ? first.message : '';
    const errorKey = message.includes('.') ? message : fallbackErrorKey;
    return { ok: false, statusCode: 400, errorKey };
  }
  return { ok: true, data: parsed.data };
}

/** Convenience for the most common case (ObjectId path param). */
export function parseObjectIdParam(
  raw: unknown,
  errorKey: string = 'common.invalidObjectId'
): PathParamResult<string> {
  return parsePathParam(raw, objectIdString(errorKey), errorKey);
}
