// @aws-ddd-api/shared/validation/zod.ts
import type { ZodType } from 'zod';

/**
 * Safely extracts Zod issues from a thrown Zod error-like value.
 *
 * @param error - Any caught error value.
 * @returns The Zod `issues` array, or an empty array when unavailable.
 */
export function getZodIssues(error: unknown): unknown[] {
  const issues = (error as { issues?: unknown })?.issues;
  return Array.isArray(issues) ? issues : [];
}

/**
 * Returns the first validation issue message from a Zod error-like value.
 *
 * @param error - Any caught error value.
 * @param fallback - Message to use when no Zod issue message exists.
 * @returns The first issue message or the fallback.
 */
export function getFirstZodIssueMessage(error: unknown, fallback = 'Invalid input'): string {
  const firstIssue = getZodIssues(error)[0] as { message?: unknown } | undefined;
  return typeof firstIssue?.message === 'string' && firstIssue.message ? firstIssue.message : fallback;
}

/**
 * Joins all validation issue messages from a Zod error-like value.
 *
 * @param error - Any caught error value.
 * @param fallback - Message to use when no Zod issue messages exist.
 * @returns A comma-separated message list or the fallback.
 */
export function getJoinedZodIssueMessages(error: unknown, fallback = 'Invalid input'): string {
  const messages = getZodIssues(error)
    .map((issue) => (issue as { message?: unknown })?.message)
    .filter((message): message is string => typeof message === 'string' && Boolean(message));

  return messages.length > 0 ? messages.join(', ') : fallback;
}

/**
 * Parses an already-decoded JSON body with a Zod schema.
 *
 * @param body - Parsed request body or other unknown input.
 * @param schema - Zod schema used to validate and type the body.
 * @returns The parsed schema output.
 * @throws The underlying Zod error when validation fails.
 */
export function parseJsonBodyWithSchema<T>(body: unknown, schema: ZodType<T>): T {
  return schema.parse(body);
}

/**
 * Successful body parsing result.
 */
export interface ParseBodySuccess<T> {
  ok: true;
  data: T;
}

/**
 * Failed body parsing result. Holds an HTTP status code and i18n error key
 * so callers can pass them straight into `response.errorResponse`.
 */
export interface ParseBodyFailure {
  ok: false;
  statusCode: number;
  errorKey: string;
}

/**
 * Discriminated union returned by `parseBody`.
 */
export type ParseBodyResult<T> = ParseBodySuccess<T> | ParseBodyFailure;

/**
 * Options that customize `parseBody` error keys and validation behavior.
 */
export interface ParseBodyOptions {
  /**
   * When true, reject bodies that are missing, non-object, or empty objects
   * with `emptyBodyErrorKey` before running the schema. Default: `false`.
   */
  requireNonEmpty?: boolean;
  /**
   * Error key returned when the raw body could not be parsed as JSON
   * (i.e. arrives at services as a string). Default: `'common.invalidBodyParams'`.
   */
  malformedJsonErrorKey?: string;
  /**
   * Error key returned when `requireNonEmpty` is true and the body is missing,
   * non-object, or an empty object. Default: `'common.missingParams'`.
   */
  emptyBodyErrorKey?: string;
  /**
   * Error key returned when no Zod issue message looks like a dotted i18n key.
   * Default: `'common.invalidBodyParams'`.
   */
  fallbackErrorKey?: string;
}

const DEFAULT_PARSE_BODY_OPTIONS: Required<ParseBodyOptions> = {
  requireNonEmpty: false,
  malformedJsonErrorKey: 'common.invalidBodyParams',
  emptyBodyErrorKey: 'common.missingParams',
  fallbackErrorKey: 'common.invalidBodyParams',
};

/**
 * Standardized request body parser for API Gateway services.
 *
 * Steps performed in order:
 * 1. If `body` is a string, the upstream JSON parse failed -> reject with
 *    `malformedJsonErrorKey`.
 * 2. If `requireNonEmpty` is set, reject missing/non-object/empty bodies with
 *    `emptyBodyErrorKey`.
 * 3. Run `schema.safeParse(body)`. On failure, return the first Zod issue
 *    message when it looks like a dotted i18n key, otherwise `fallbackErrorKey`.
 *
 * Callers can inline the failure result into `response.errorResponse`:
 *
 * ```ts
 * const result = parseBody(ctx.body, schema);
 * if (!result.ok) {
 *   return response.errorResponse(result.statusCode, result.errorKey, ctx.event);
 * }
 * const data = result.data;
 * ```
 *
 * @param body - Decoded request body from the route context.
 * @param schema - Zod schema describing the expected body shape.
 * @param options - Optional overrides for error keys and emptiness checks.
 * @returns Discriminated union with parsed data or HTTP error key.
 */
export function parseBody<T>(
  body: unknown,
  schema: ZodType<T>,
  options: ParseBodyOptions = {}
): ParseBodyResult<T> {
  const {
    requireNonEmpty,
    malformedJsonErrorKey,
    emptyBodyErrorKey,
    fallbackErrorKey,
  } = { ...DEFAULT_PARSE_BODY_OPTIONS, ...options };

  if (typeof body === 'string') {
    return { ok: false, statusCode: 400, errorKey: malformedJsonErrorKey };
  }

  if (requireNonEmpty) {
    if (!body || typeof body !== 'object' || Array.isArray(body)) {
      return { ok: false, statusCode: 400, errorKey: emptyBodyErrorKey };
    }
    if (Object.keys(body as Record<string, unknown>).length === 0) {
      return { ok: false, statusCode: 400, errorKey: emptyBodyErrorKey };
    }
  }

  const parsed = schema.safeParse(body);
  if (!parsed.success) {
    const message = getFirstZodIssueMessage(parsed.error, fallbackErrorKey);
    const errorKey = message.includes('.') ? message : fallbackErrorKey;
    return { ok: false, statusCode: 400, errorKey };
  }

  return { ok: true, data: parsed.data };
}
