// @aws-ddd-api/shared/validation/zod.ts
import type { ZodType } from 'zod';
import type { APIGatewayProxyEvent } from 'aws-lambda';
import multipart from 'lambda-multipart-parser';

/**
 * Safely extracts Zod issues from a thrown Zod error-like value.
 *
 * @param error - Any caught error value.
 * @returns The Zod `issues` array, or an empty array when unavailable.
 */
export function getZodIssues(error: unknown): unknown[] {
  const issues = (error as { issues?: unknown })?.issues;
  return Array.isArray(issues) ? issues : [];
}

/**
 * Returns the first validation issue message from a Zod error-like value.
 *
 * @param error - Any caught error value.
 * @param fallback - Message to use when no Zod issue message exists.
 * @returns The first issue message or the fallback.
 */
export function getFirstZodIssueMessage(error: unknown, fallback = 'Invalid input'): string {
  const firstIssue = getZodIssues(error)[0] as { message?: unknown } | undefined;
  return typeof firstIssue?.message === 'string' && firstIssue.message ? firstIssue.message : fallback;
}

/**
 * Joins all validation issue messages from a Zod error-like value.
 *
 * @param error - Any caught error value.
 * @param fallback - Message to use when no Zod issue messages exist.
 * @returns A comma-separated message list or the fallback.
 */
export function getJoinedZodIssueMessages(error: unknown, fallback = 'Invalid input'): string {
  const messages = getZodIssues(error)
    .map((issue) => (issue as { message?: unknown })?.message)
    .filter((message): message is string => typeof message === 'string' && Boolean(message));

  return messages.length > 0 ? messages.join(', ') : fallback;
}

/**
 * Parses an already-decoded JSON body with a Zod schema.
 *
 * @param body - Parsed request body or other unknown input.
 * @param schema - Zod schema used to validate and type the body.
 * @returns The parsed schema output.
 * @throws The underlying Zod error when validation fails.
 */
export function parseJsonBodyWithSchema<T>(body: unknown, schema: ZodType<T>): T {
  return schema.parse(body);
}

/**
 * Successful body parsing result.
 */
export interface ParseBodySuccess<T> {
  ok: true;
  data: T;
}

/**
 * Failed body parsing result. Holds an HTTP status code and i18n error key
 * so callers can pass them straight into `response.errorResponse`.
 */
export interface ParseBodyFailure {
  ok: false;
  statusCode: number;
  errorKey: string;
}

/**
 * Discriminated union returned by `parseBody`.
 */
export type ParseBodyResult<T> = ParseBodySuccess<T> | ParseBodyFailure;

/**
 * Options that customize `parseBody` error keys and validation behavior.
 */
export interface ParseBodyOptions {
  /**
   * When true, reject bodies that are missing, non-object, or empty objects
   * with `emptyBodyErrorKey` before running the schema. Default: `true`.
   */
  requireNonEmpty?: boolean;
  /**
   * Error key returned when the raw body could not be parsed as JSON
   * (i.e. arrives at services as a string). Default: `'common.invalidBodyParams'`.
   */
  malformedJsonErrorKey?: string;
  /**
   * Error key returned when `requireNonEmpty` is true and the body is missing,
   * non-object, or an empty object. Default: `'common.missingParams'`.
   */
  emptyBodyErrorKey?: string;
  /**
   * Error key returned when no Zod issue message looks like a dotted i18n key.
   * Default: `'common.invalidBodyParams'`.
   */
  fallbackErrorKey?: string;
}

const DEFAULT_PARSE_BODY_OPTIONS: Required<ParseBodyOptions> = {
  requireNonEmpty: true,
  malformedJsonErrorKey: 'common.invalidBodyParams',
  emptyBodyErrorKey: 'common.missingBodyParams',
  fallbackErrorKey: 'common.invalidBodyParams',
};

/**
 * Standardized request body parser for API Gateway services.
 *
 * Steps performed in order:
 * 1. If `body` is a string, the upstream JSON parse failed -> reject with
 *    `malformedJsonErrorKey`.
 * 2. If `requireNonEmpty` is set, reject missing/non-object/empty bodies with
 *    `emptyBodyErrorKey`.
 * 3. Run `schema.safeParse(body)`. On failure, return the first Zod issue
 *    message when it looks like a dotted i18n key, otherwise `fallbackErrorKey`.
 *
 * Callers can inline the failure result into `response.errorResponse`:
 *
 * ```ts
 * const result = parseBody(ctx.body, schema);
 * if (!result.ok) {
 *   return response.errorResponse(result.statusCode, result.errorKey, ctx.event);
 * }
 * const data = result.data;
 * ```
 *
 * @param body - Decoded request body from the route context.
 * @param schema - Zod schema describing the expected body shape.
 * @param options - Optional overrides for error keys and emptiness checks.
 * @returns Discriminated union with parsed data or HTTP error key.
 */
export function parseBody<T>(
  body: unknown,
  schema: ZodType<T>,
  options: ParseBodyOptions = {}
): ParseBodyResult<T> {
  const {
    requireNonEmpty,
    malformedJsonErrorKey,
    emptyBodyErrorKey,
    fallbackErrorKey,
  } = { ...DEFAULT_PARSE_BODY_OPTIONS, ...options };

  if (typeof body === 'string') {
    return { ok: false, statusCode: 400, errorKey: malformedJsonErrorKey };
  }

  if (requireNonEmpty) {
    if (!body || typeof body !== 'object' || Array.isArray(body)) {
      return { ok: false, statusCode: 400, errorKey: emptyBodyErrorKey };
    }
    if (Object.keys(body as Record<string, unknown>).length === 0) {
      return { ok: false, statusCode: 400, errorKey: emptyBodyErrorKey };
    }
  }

  const parsed = schema.safeParse(body);
  if (!parsed.success) {
    const message = getFirstZodIssueMessage(parsed.error, fallbackErrorKey);
    const errorKey = message.includes('.') ? message : fallbackErrorKey;
    return { ok: false, statusCode: 400, errorKey };
  }

  return { ok: true, data: parsed.data };
}

/**
 * A single file entry from a parsed multipart/form-data request.
 */
export interface ParsedMultipartFile {
  content?: Buffer;
  filename?: string;
  contentType?: string;
  fieldname?: string;
}

/**
 * Successful multipart body parsing result.
 */
export interface ParseMultipartBodySuccess<T> {
  ok: true;
  data: T;
  files: ParsedMultipartFile[];
}

/**
 * Discriminated union returned by `parseMultipartBody`.
 */
export type ParseMultipartBodyResult<T> = ParseMultipartBodySuccess<T> | ParseBodyFailure;

/**
 * Options that customize `parseMultipartBody` error keys and validation behavior.
 */
export interface ParseMultipartBodyOptions {
  /**
   * Called with raw fields before normalization. Return an i18n error key to
   * short-circuit with a 400, or `null` to continue. Use this for allowlist /
   * pre-Zod field checks that need to return a specific error key.
   */
  validate?: (rawFields: Record<string, unknown>) => string | null;
  /**
   * Transforms raw string fields before Zod validation (e.g. coerce booleans
   * and numbers that arrive as strings from lambda-multipart-parser).
   */
  normalize?: (rawFields: Record<string, unknown>) => Record<string, unknown>;
  /** Error key when the multipart body cannot be parsed. Default: `'common.invalidBodyParams'`. */
  parseErrorKey?: string;
  /** Error key when no Zod issue message looks like a dotted i18n key. Default: `'common.invalidBodyParams'`. */
  fallbackErrorKey?: string;
}

/**
 * Standardized multipart/form-data parser for API Gateway services.
 *
 * Steps performed in order:
 * 1. Call `multipart.parse(event)`. On failure return `parseErrorKey`.
 * 2. Separate `files` from scalar fields.
 * 3. If `validate` is provided, call it with raw fields. A non-null return is
 *    an i18n error key → reject with 400.
 * 4. If `normalize` is provided, apply it to the raw fields.
 * 5. Run `schema.safeParse(fields)` with the same i18n-key error handling as
 *    `parseBody`.
 *
 * On success returns `{ ok: true, data, files }` so callers can access both
 * the validated payload and uploaded file buffers.
 *
 * ```ts
 * const result = await parseMultipartBody(ctx.event, schema, { normalize });
 * if (!result.ok) {
 *   return response.errorResponse(result.statusCode, result.errorKey, ctx.event);
 * }
 * const { data, files } = result;
 * ```
 */
export async function parseMultipartBody<T>(
  event: APIGatewayProxyEvent,
  schema: ZodType<T>,
  options: ParseMultipartBodyOptions = {}
): Promise<ParseMultipartBodyResult<T>> {
  const {
    validate,
    normalize,
    parseErrorKey = 'common.invalidBodyParams',
    fallbackErrorKey = 'common.invalidBodyParams',
  } = options;

  let form: Record<string, unknown>;
  try {
    form = (await multipart.parse(event)) as Record<string, unknown>;
  } catch {
    return { ok: false, statusCode: 400, errorKey: parseErrorKey };
  }

  const files = (form['files'] ?? []) as ParsedMultipartFile[];
  const { files: _files, ...rawFields } = form;

  if (validate) {
    const errorKey = validate(rawFields);
    if (errorKey !== null) {
      return { ok: false, statusCode: 400, errorKey };
    }
  }

  const fields = normalize ? normalize(rawFields) : rawFields;

  const parsed = schema.safeParse(fields);
  if (!parsed.success) {
    const message = getFirstZodIssueMessage(parsed.error, fallbackErrorKey);
    const errorKey = message.includes('.') ? message : fallbackErrorKey;
    return { ok: false, statusCode: 400, errorKey };
  }

  return { ok: true, data: parsed.data, files };
}
